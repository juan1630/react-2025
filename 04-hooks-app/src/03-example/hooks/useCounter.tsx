import { useState } from "react";

export const useCounter = (initialValue: number = 1) => {
  const [counter, setCounter] = useState<number>(initialValue);

  const increment = () => {
    setCounter((prev) => prev + 1);
  };

  const decrement = () => {
    setCounter((prev) => {
      if (prev === 0) return 0;
      return prev - 1;
    });
  };

  return {
    counter,
    increment,
    decrement,
  };
};
