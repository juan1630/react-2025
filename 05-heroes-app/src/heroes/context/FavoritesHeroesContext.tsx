import {
  createContext,
  useEffect,
  useState,
  type PropsWithChildren,
} from "react";
import type { Hero } from "../interfaces/hero.interface";

interface FavoritesHeroesContext {
  //State
  favorites: Hero[];
  favoriteCount: number;

  //Methods
  isFavorite: (hero: Hero) => boolean;
  toggleFavorite: (hero: Hero) => void;
}

export const FavoritesHeroesContext = createContext(
  {} as FavoritesHeroesContext
);

const getHeroesFromLocalStorage = (): Hero[] => {
  const favorites = localStorage.getItem("favorites");
  return favorites ? JSON.parse(favorites) : [];
};

export const FavoriteHeroProvider = ({ children }: PropsWithChildren) => {
  const [favorites, setFavorites] = useState<Hero[]>(
    getHeroesFromLocalStorage()
  );

  useEffect(() => {
    localStorage.setItem("favorites", JSON.stringify(favorites));
  }, [favorites]);

  const toggleFavorite = (hero: Hero) => {
    const heroExists = favorites.find((heroItem) => heroItem.id === hero.id);
    if (heroExists) {
      const newHeroesFevorites = favorites.filter(
        (prevHero) => prevHero.id !== hero.id
      );
      setFavorites(newHeroesFevorites);
      return;
    }

    setFavorites((prev) => [...prev, hero]);
  };

  return (
    <FavoritesHeroesContext.Provider
      value={{
        favoriteCount: favorites.length,
        favorites,
        isFavorite: (hero: Hero) =>
          favorites.some((heroItem) => heroItem.id === hero.id),
        toggleFavorite: toggleFavorite,
      }}
    >
      {children}
    </FavoritesHeroesContext.Provider>
  );
};
