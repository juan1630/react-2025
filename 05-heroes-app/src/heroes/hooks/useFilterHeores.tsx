import { useQuery } from "@tanstack/react-query";
import { searchHeroesAction } from "../actions/get-hero-by-filters.action";

export const useFilterHeores = (name: string = "") => {
  return  useQuery({
    queryKey: ["search",  name],
    queryFn: () => searchHeroesAction({ name}),
    staleTime: 1000 * 60 * 5, //5 minutos
  });
};
