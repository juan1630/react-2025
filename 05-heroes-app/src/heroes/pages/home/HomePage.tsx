import { use, useMemo, useState } from "react";
import { useSearchParams } from "react-router";
import { CustomJumBotron } from "@/components/ui/Custom/CustomJumBotron";
import { HeroStats } from "../hero/Components/HeroStats";
import { CustomBreadCrumbs } from "@/components/ui/Custom/CustomBreadCrumbs";
import { Filter, Plus, Search } from "lucide-react";
import { Input } from "@/components/ui/input";
import {
  Dialog,
  DialogHeader,
  DialogTrigger,
  DialogTitle,
  DialogDescription,
  DialogContent,
} from "@/components/ui/dialog";
import { Button } from "@/components/ui/button";
import { Label } from "@/components/ui/label";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
import { HeroGrid } from "../hero/Components/HeroGrid";
import { Badge } from "@/components/ui/badge";
import { CustomPagination } from "@/components/ui/Custom/CustomPagination";
import { useHeroSummary } from "../../hooks/useHeroSummary";
import { usePaginatedHero } from "@/heroes/hooks/usePaginatedHero";
import { FavoritesHeroesContext } from "@/heroes/context/FavoritesHeroesContext";

export const HomePage = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const { favoriteCount, favorites } = use(FavoritesHeroesContext);

  const tab = searchParams.get("tab") ?? "all";
  const page = searchParams.get("page") ?? "1";
  const limit = searchParams.get("limit") ?? "6";
  const category = searchParams.get("category") ?? "all";
  

  const selectedTabs = useMemo(() => {
    const validTabs = ["all", "favorites", "heroes", "villains"];
    return validTabs.includes(tab) ? tab : "all";
  }, [tab]);

  const { data: summary } = useHeroSummary();
  const { data: heroesResponse } = usePaginatedHero(+page, +limit, category);

  const [searchTerm, setSearchTerm] = useState("");
  const [isDialogOpen, setIsDialogOpen] = useState(false);

  return (
    <>
      {/* Header */}
      <CustomJumBotron
        title="Universo de SuperHéroes"
        description="Descubre, explora y administra super héroes y villanos"
      />

      <CustomBreadCrumbs currentPage="heroes" />

      {/* Stats */}
      <HeroStats />

      {/* Search and Add Hero Section */}
      <div className="flex flex-col sm:flex-row gap-4 mb-8">
        <div className="relative flex-1">
          <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-4 w-4" />
          <Input
            placeholder="Search heroes by name, alias, powers, or team..."
            value={searchTerm}
            onChange={(e) => setSearchTerm(e.target.value)}
            className="pl-10"
          />
        </div>
        <Dialog open={isDialogOpen} onOpenChange={setIsDialogOpen}>
          <DialogTrigger asChild>
            <Button className="flex items-center gap-2">
              <Plus className="h-4 w-4" />
              Add Hero
            </Button>
          </DialogTrigger>
          <DialogContent className="max-w-md">
            <DialogHeader>
              <DialogTitle>Add New Hero</DialogTitle>
              <DialogDescription>
                Create a new superhero profile with their details and abilities.
              </DialogDescription>
            </DialogHeader>
            <div className="grid gap-4 py-4">
              <div className="grid gap-2">
                <Label htmlFor="alias">Hero Alias</Label>
              </div>
              <div className="grid gap-2">
                <Label htmlFor="powers">Powers (comma-separated)</Label>
              </div>
            </div>
          </DialogContent>
        </Dialog>
      </div>

      {/* Results Count */}

      {/* Tabs */}
      <Tabs value={selectedTabs} className="mb-8">
        <TabsList className="grid w-full grid-cols-4">
          <TabsTrigger
            onClick={() =>
              setSearchParams((prev) => {
                prev.set("tab", "all");
                prev.set("category", "all");
                prev.set("page", "1");
                return prev;
              })
            }
            value="all"
          >
            All Characters ( {summary?.totalHeroes} )
          </TabsTrigger>

          <TabsTrigger
            onClick={() =>
              setSearchParams((prev) => {
                prev.set("tab", "favorites");
                return prev;
              })
            }
            value="favorites"
          >
            Favorites ({favoriteCount})
          </TabsTrigger>
          <TabsTrigger
            onClick={() =>
              setSearchParams((prev) => {
                prev.set("tab", "heores");
                prev.set("category", "hero");
                prev.set("page", "1");
                return prev;
              })
            }
            value="heroes"
          >
            Heroes ( {summary?.heroCount})
          </TabsTrigger>
          <TabsTrigger
            value="villains"
            onClick={() =>
              setSearchParams((prev) => {
                prev.set("tab", "villains");
                prev.set("category", "villain");
                prev.set("page", "1");
                return prev;
              })
            }
          >
            Villains ({summary?.villainCount})
          </TabsTrigger>
        </TabsList>
        <TabsContent value="all">
          <h1 className="text-3xl text-center mb-5">Todos los personajes</h1>
          <HeroGrid heroes={heroesResponse?.heroes ?? []} />
        </TabsContent>
        <TabsContent value="favorites">
          <h1>Personajes favoritos</h1>
          <HeroGrid heroes={favorites} />
        </TabsContent>
        <TabsContent value="heroes">
          <h1>Todos los héroes</h1>
          <HeroGrid heroes={heroesResponse?.heroes ?? []} />
        </TabsContent>
        <TabsContent value="villains">
          <h1>Todos los villanos</h1>
          <HeroGrid heroes={heroesResponse?.heroes ?? []} />
        </TabsContent>
      </Tabs>

      {/* Results info */}
      <div className="flex justify-between items-center mb-6">
        <div className="flex items-center gap-4">
          <p className="text-gray-600">Showing 6 of 16 characters</p>
          <Badge variant="secondary" className="flex items-center gap-1">
            <Filter className="h-3 w-3" />
            Filtered
          </Badge>
        </div>
      </div>

      {selectedTabs !== "favorites" ? (
        <CustomPagination totalPages={heroesResponse?.pages ?? 1} />
      ) : null}
    </>
  );
};
