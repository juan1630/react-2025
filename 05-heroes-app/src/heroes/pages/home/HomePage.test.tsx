import { fireEvent, render, screen } from "@testing-library/react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import { HomePage } from "./HomePage";
import { MemoryRouter } from "react-router";
import { usePaginatedHero } from "@/heroes/hooks/usePaginatedHero";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { FavoriteHeroProvider } from "@/heroes/context/FavoritesHeroesContext";


vi.mock("@/heroes/hooks/usePaginatedHero");
const mockUsePaginatedHero = vi.mocked(usePaginatedHero);

mockUsePaginatedHero.mockReturnValue({
  data: [],
  isLoading: false,
  isError: false,
} as unknown as ReturnType<typeof mockUsePaginatedHero>);

const queryClient = new QueryClient();

const renderHomePage = (initialEntries: string[] = ["/"]) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <QueryClientProvider client={queryClient}>
        <FavoriteHeroProvider>
        <HomePage />
        </FavoriteHeroProvider>
      </QueryClientProvider>
    </MemoryRouter>
  );
};

describe("HomePage", () => {

    beforeEach(()=> {vi.clearAllMocks()})

  test("should render HomePage with default values", () => {
    const { container } = renderHomePage();
    expect(container).toMatchSnapshot();
  });

  test('should call usePaginated with default values', ()=> {
    renderHomePage()

    expect(mockUsePaginatedHero).toHaveBeenCalled()
    expect(mockUsePaginatedHero).toHaveBeenCalledWith(1,6,'all')
  })

  test('should call usePaginated with custome params', ()=> {
    renderHomePage(['/?page=2&limit=10&category=villains'])

    expect(mockUsePaginatedHero).toHaveBeenCalled()
    expect(mockUsePaginatedHero).toHaveBeenCalledWith(2,10,'villains')
  })

  test('should call usePaginatedHero with  default page and same limit o tab click', ()=> {
    renderHomePage(['/?tab=favorites&page=2&limit=10'])
    const [ ,,,villainsTab ] = screen.getAllByRole('tab')
    
    fireEvent.click(villainsTab)
    expect(mockUsePaginatedHero).toHaveBeenCalled()
    expect(mockUsePaginatedHero).toHaveBeenCalledWith(1,10, 'villain')
  })
});
