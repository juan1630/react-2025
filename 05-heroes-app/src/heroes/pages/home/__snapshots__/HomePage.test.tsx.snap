// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage > should render HomePage with default values 1`] = `
<div>
  <div
    class="text-center mb-8"
  >
    <h1
      class="text-4xl font-bold text-gray-900 mb-2"
    >
      Universo de SuperHéroes
    </h1>
    <p
      class="text-gray-600"
    >
      Descubre, explora y administra super héroes y villanos
    </p>
  </div>
  <nav
    aria-label="breadcrumb"
    class="mb-4"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
           Inicio 
        </a>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <li
          aria-hidden="true"
          class="[&>svg]:size-3.5"
          data-slot="breadcrumb-separator"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-slash"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 2 2 22"
            />
          </svg>
        </li>
        <a
          class="hover:text-foreground transition-colors text-black"
          data-slot="breadcrumb-link"
        >
          heroes
           
        </a>
      </li>
    </ol>
  </nav>
  <div
    class="grid grid-cols-2 md:grid-cols-4 gap-4 mb-8"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Total de personajes
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-users h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <path
            d="M16 3.128a4 4 0 0 1 0 7.744"
          />
          <path
            d="M22 21v-2a4 4 0 0 0-3-3.87"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold"
        >
           
        </div>
        <div
          class="flex gap-1 mt-2"
        >
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 text-xs"
            data-slot="badge"
          >
             Heroes
          </span>
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 text-xs"
            data-slot="badge"
          >
             Villains
          </span>
        </div>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Favoritos
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-heart h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold text-red-600"
          data-testid="favorite-count"
        >
          0
        </div>
        <p
          class="text-xs text-muted-foreground"
          data-testid="favorite-percentage"
        >
           % of total
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Fuerte
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-zap h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 14a1 1 0 0 1-.78-1.63l9.9-10.2a.5.5 0 0 1 .86.46l-1.92 6.02A1 1 0 0 0 13 10h7a1 1 0 0 1 .78 1.63l-9.9 10.2a.5.5 0 0 1-.86-.46l1.92-6.02A1 1 0 0 0 11 14z"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        />
        <p
          class="text-xs text-muted-foreground"
        >
          Strength: 
          /10
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Inteligente
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-trophy h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 14.66v1.626a2 2 0 0 1-.976 1.696A5 5 0 0 0 7 21.978"
          />
          <path
            d="M14 14.66v1.626a2 2 0 0 0 .976 1.696A5 5 0 0 1 17 21.978"
          />
          <path
            d="M18 9h1.5a1 1 0 0 0 0-5H18"
          />
          <path
            d="M4 22h16"
          />
          <path
            d="M6 9a6 6 0 0 0 12 0V3a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1z"
          />
          <path
            d="M6 9H4.5a1 1 0 0 1 0-5H6"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        />
        <p
          class="text-xs text-muted-foreground"
        >
          Intelligence: 
           /10
        </p>
      </div>
    </div>
  </div>
  <div
    class="flex flex-col sm:flex-row gap-4 mb-8"
  >
    <div
      class="relative flex-1"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive pl-10"
        data-slot="input"
        placeholder="Search heroes by name, alias, powers, or team..."
        value=""
      />
    </div>
    <button
      aria-controls="radix-«r0»"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="cursor-pointer justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 flex items-center gap-2"
      data-slot="dialog-trigger"
      data-state="closed"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
      Add Hero
    </button>
  </div>
  <div
    class="flex flex-col gap-2 mb-8"
    data-orientation="horizontal"
    data-slot="tabs"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="bg-muted text-muted-foreground h-9 items-center justify-center rounded-lg p-[3px] grid w-full grid-cols-4"
      data-orientation="horizontal"
      data-slot="tabs-list"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-«r3»-content-all"
        aria-selected="true"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="active"
        id="radix-«r3»-trigger-all"
        role="tab"
        tabindex="-1"
        type="button"
      >
        All Characters ( 
         )
      </button>
      <button
        aria-controls="radix-«r3»-content-favorites"
        aria-selected="false"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="inactive"
        id="radix-«r3»-trigger-favorites"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Favorites (
        0
        )
      </button>
      <button
        aria-controls="radix-«r3»-content-heroes"
        aria-selected="false"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="inactive"
        id="radix-«r3»-trigger-heroes"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Heroes ( 
        )
      </button>
      <button
        aria-controls="radix-«r3»-content-villains"
        aria-selected="false"
        class="data-[state=active]:bg-background dark:data-[state=active]:text-foreground focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:outline-ring dark:data-[state=active]:border-input dark:data-[state=active]:bg-input/30 text-foreground dark:text-muted-foreground inline-flex h-[calc(100%-1px)] flex-1 items-center justify-center gap-1.5 rounded-md border border-transparent px-2 py-1 text-sm font-medium whitespace-nowrap transition-[color,box-shadow] focus-visible:ring-[3px] focus-visible:outline-1 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:shadow-sm [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-slot="tabs-trigger"
        data-state="inactive"
        id="radix-«r3»-trigger-villains"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Villains (
        )
      </button>
    </div>
    <div
      aria-labelledby="radix-«r3»-trigger-all"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="active"
      id="radix-«r3»-content-all"
      role="tabpanel"
      style="animation-duration: 0s;"
      tabindex="0"
    >
      <h1
        class="text-3xl text-center mb-5"
      >
        Todos los personajes
      </h1>
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-8 mb-8"
      />
    </div>
    <div
      aria-labelledby="radix-«r3»-trigger-favorites"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="inactive"
      hidden=""
      id="radix-«r3»-content-favorites"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="radix-«r3»-trigger-heroes"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="inactive"
      hidden=""
      id="radix-«r3»-content-heroes"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="radix-«r3»-trigger-villains"
      class="flex-1 outline-none"
      data-orientation="horizontal"
      data-slot="tabs-content"
      data-state="inactive"
      hidden=""
      id="radix-«r3»-content-villains"
      role="tabpanel"
      tabindex="0"
    />
  </div>
  <div
    class="flex justify-between items-center mb-6"
  >
    <div
      class="flex items-center gap-4"
    >
      <p
        class="text-gray-600"
      >
        Showing 6 of 16 characters
      </p>
      <span
        class="justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 flex items-center gap-1"
        data-slot="badge"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-funnel h-3 w-3"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 20a1 1 0 0 0 .553.895l2 1A1 1 0 0 0 14 21v-7a2 2 0 0 1 .517-1.341L21.74 4.67A1 1 0 0 0 21 3H3a1 1 0 0 0-.742 1.67l7.225 7.989A2 2 0 0 1 10 14z"
          />
        </svg>
        Filtered
      </span>
    </div>
  </div>
  <div
    class="flex items-center justify-center space-x-2"
  >
    <button
      class="cursor-pointer inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
      data-slot="button"
      disabled=""
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-left h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m15 18-6-6 6-6"
        />
      </svg>
      Anteriores
    </button>
    <button
      class="cursor-pointer inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
      data-slot="button"
    >
      1
    </button>
    <button
      class="cursor-pointer inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
      data-slot="button"
      disabled=""
    >
      Siguientes
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-right h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m9 18 6-6-6-6"
        />
      </svg>
    </button>
  </div>
</div>
`;
